### ISSUE TITLE
Support using index in backfilling

### ISSUE THREAD

**Issue by hzxa21** [risingwavelabs/risingwave#19298]
### Is your feature request related to a problem? Please describe.

> basically if we are creating a sink from A to B, and we know B or B's downstream prefer the backfilling rows to be ordered by X, we better add ORDER BY X to A if A is a MV, or create a new MV A' as SELECT .. from A ORDER BY X and create the sink from A'

Sometimes we need backfilling to emit rows in a specific ordering to reduce unnecessary downstream changes. Currently we can work around it by creating a MV with ORDER BY and use this MV for backfilling but this MV cannot be dropped after backfilling because it exists in the streaming graph. Instead, using index for backfilling seems more flexible.

**Comment by xxchan** [comment]
This was also once mentioned here https://github.com/risingwavelabs/risingwave/issues/12824#issuecomment-1776519122

**Comment by kwannoel** [comment]
> Currently we can work around it by creating a MV with ORDER BY and use this MV for backfilling but this MV cannot be dropped after backfilling because it exists in the streaming graph. Instead, using index for backfilling seems more flexible.

Do you mean that you'd like to use the index for backfilling side, but the MV for the upstream side? And then be able to drop the index after backfilling done?

**Comment by xxchan** [comment]
Yes, the index is only useful for backfilling. Persisting data twice both in MV and index seems to wasteful.

**Comment by BugenZhao** [comment]
> Persisting data twice both in MV and index seems to wasteful.

That sounds a bit arbitrary to me. Clustered index is essentially a trade-off between time and space, and may not be labeled "wasteful". 🤣 If a user plans to create multiple MVs that can be accelerated with this index, maintaining it (compared to make it "temporary" for a specific backfilling MV) can be highly beneficial.

**Comment by BugenZhao** [comment]
> Do you mean that you'd like to use the index for backfilling side, but the MV for the upstream side? And then be able to drop the index after backfilling done?

Regarding implementation, I think the key point is to allow removing the reference to the table to be scanned once the backfill is complete. Whether the table is an index or a materialized view with `ORDER BY` does not make much difference.

**Comment by kwannoel** [comment]
Lower priority. Users may rely on index not just during backfill, but during streaming phase, since they index can control stream distribution after #19321.

They can already create stream jobs directly on index for this usecase.
